cli/azd/pkg/tools/git/git.go

```diff
@@ -251,6 +253,33 @@ func (cli *gitCli) IsUntrackedFile(ctx context.Context, repositoryPath string, f
 	return false, nil
 }
 
+// SetAzdoPatAuth creates git config for the repositoryPath like
+//
+// [credential]
+// helper =
+// helper = azdoPatFile
+//
+// This way, git commands run from repositoryPath will use $AZURE_DEVOPS_EXT_PAT for auth.
+// Note: Removes any previous configuration for the credential.
+// Note: `helper = ` is intentional to break the chain of previously configured global helpers.
+func (cli *gitCli) SetAzdoPatAuth(ctx context.Context, repositoryPath, azdoPatFile string) error {
+	runArgs := newRunArgs(
+		"-C", repositoryPath,
+		"config", "--local", "--replace-all", "credential.helper", "")
+	if _, err := cli.commandRunner.Run(ctx, runArgs); err != nil {
+		return fmt.Errorf("failed to set break for azdo credential helper: %w", err)
+	}
+
+	runArgs = newRunArgs(
```

Comment:
> If $AZURE_DEVOPS_EXT_PAT is empty or invalid, what happens when I try to git push to the repository and it requires authentication?

azdo will always require auth for https (using PAT for password).  `git push` alone would fail if the pat is not set within the environment.  I was thinking it was not too bad to tell folks to set the PAT to make it easier, but I also understand that's not what we really want azd to do.

Can we go with this approach and wait for customer's feedback?  

From my experience hearing from folks using azdo, setting up the local repo is a pain, and this is at least one step to make this easier.  As azd is taking care of git-credential.

For CodeSpaces.  GitHub automatically set a global credential helper which makes any `git push` to azdo to fail.  So, even without azd doing anything, using git for azdo from a codespace is already challenging (folks needs to understand about git-credentials and config).

I would like to wait to hear from customers and then make changes based on that.

Reply:
`azd` should not be controlling (especially when it comes to destructive actions) git configuration for the repository. What if the user had a perfectly well setup git credential helper configuration? This leads to user data loss. At minimum, a consent prompt is necessary. Please correct my understanding if I'm wrong here.

For Codespaces, if Codespaces has set up git machine-level configuration that is unfriendly to all git users, perhaps we need an issue logged there?

The secondary question is then can we do this ephemerally without modifying user data? I haven't seen PM input here on the criticalness of the issue, but let's try to think clearly about the impact here.

Reply:
> What if the user had a perfectly well setup git credential helper configuration? This leads to user data loss.

IMHO, for a user with a well git-credential config, it would be trivial to understand what is happening and provide feedback to us about it. Which is what I would hope for.
But, for users with no git-credential config, we would keep receiving feedback about why is it that hard to push changes to azdo, or why does azd keeps asking for PAT on every operation.

So, my idea here is to invest on making/creating an easy flow specially for beginners. 

Also, we would be setting the configuration locally (only for the repo).  
The expectation is still to be able to demonstrate azd's flow from `azd init` to `azd pipeline config`, where adding git credential helpers manually in the middle of the process is not something expected.

> For Codespaces, if Codespaces has set up git machine-level configuration that is unfriendly to all git users, perhaps we need an issue logged there?

I don't know..  I think they do it b/c there's is an expectation that a codespace is for opening a GitHub repository and working on it.  Not really as a general dev machine (like cloning azdo repos from the codespace, or else)

> The secondary question is then can we do this ephemerally without modifying user data?

I think so, but I am dreaming on this idea:  If azd is telling customers they can set the PAT as env var so it is not asked all the times, then azd can do a great job to provide a first class experience there :) 


I can update the credential helper to only work if the ENV VAR is set, otherwise ignore the helper.. I think that's an easy fix.

Reply:
> If $AZURE_DEVOPS_EXT_PAT is empty or invalid, what happens when I try to git push to the repository and it requires authentication?

@ellismg , I have pushed one more commit to make the credential-helper to only produce a password if AZURE_DEVOPS_EXT_PAT is set with a value..  otherwise it won't output anything and git will ignore it and fall back to prompt for the PAT, like:
![image](https://user-images.githubusercontent.com/24213737/235814445-02813628-53e3-4499-9ff9-accaada5c1e5.png)

By doing this, providing PAT to azd will give a smoothly experience e2e, and if the PAT is not set in system env, the git config is not affected.

The only corner case if what @weikanglim mentioned...  in case someone has created a local credential helper for the repo and one day decides to run `azd pipeline config --provider azdo` and see its credential config be replaced :).  But I wish we could wait until that ever happens...  I really doubt it

Reply:
It feels like the problem we are solving here is "it's hard to connect to a private repository on AzDo" and we feel obligated to fix this because the `pipeline config` flow pushes you towards creating a private repository when targeting `azdo`.  It's not a problem for github because as part of `gh auth login` they set up credential helpers and generate PATs and all that jazz for you.

From the linked issue, it sounds like this was only a problem on CodeSpaces and that's only because CodeSpaces is already configured to use a different credential helper?  Why does this work correctly other places?

I guess like @weikanglim I'm wondering what the severity of the problem is.  If this really is a problem and there's not good existing solution, I'd rather make this more first class in `azd`  and have an explicit gesture for "hook up a credential provider in this repository that spits back a PAT" and guide the user towards that gesture.

I do think this is a useful workaround - perhaps we could document it in the linked issue for now and ask folks to thumbs up the comment or something if they ended up using it?

Reply:
> t's not a problem for github because as part of gh auth login they set up credential helpers and generate PATs and all that jazz for you.

Sort of.  For GitHub, we have support for `ssh`.  Then, users would typically set up a key for GitHub and `git` will automatically try to use the `~/.ssh` key for accessing the repo.
But if the repo is `https`, git will still prompt users for credentials.  On windows, `git for windows` comes with a credential which allows you to open a mini browser to login to github. That creates a PAT and the credential handles that PAT.
If you are using the gh-cli.  The login will create a PAT for the gh-cli and that is handled by gh-cli, but not automatically shared to `git`.  You need to run `gh auth setup-git` and gh-cli will create a global config for git to use gh-cli as credential helper for GitHub repos.

What we do for GitHub is to call `gitCli.SetGitHubAuthForRepo()` after we login to `gh cli` and we create the credential helper locally for the repo only to use the gh-cli to authenticate to GitHub if the repo is using HTTPS.

I followed the same strategy for azdo, but there's not azdo-cli.  So, the less sophisticated solution here is to use that bash script instead of another binary that handles a PAT and make it just a bridge to an env-var.  If the env-var is not set, the credential-helper is ignored and git will just prompt user for auth.

For codespaces, the issue is more complex, as there is a global configuration for git with a credential-helper and also with `GIT_ASKPASS` (to use VSCode when prompting for input values).
We need to cut the wire for our repo to use the credential helper when using azdo (by adding a local credential helper like `helper =`).  This will prevent git from using the global config and going directly to prompt for auth.


> I do think this is a useful workaround - perhaps we could document it in the linked issue for now and ask folks to thumbs up the comment or something if they ended up using it?

From my previous experience with `azd auth login` on codespaces, the team is usually against using documented workarounds like this to make things work (like the login with `curl`, or the url-replace).
We would be telling about users something like:
    `azd pipeline config` will fail to push your code to azdo until you set up git to use your PAT.  Follow this steps to make this....  
So, instead, I am trying to do as much as we can with azd to mitigate the context switching, the manual tool setup, etc etc.  That's my believe.

On the other hand, I see also a benefit of moving these things as user-responsibility and provide docs and links about how to set things up.  That is less things to maintain / keep working :).  That being said, if adding docs about how to make things work is acceptable here, I am happy to go on that direction.  But I am 100% confident that I will see a future request to make azd work without those manual steps XD  (demos are not cool when you need to extra steps, easy to miss/forget and which are the key for a successful demo)

Reply:
Stepping back from `azd` for a second, I am trying to think about how someone who was just using codepsaces and AzDo would be expected to work.

You mention there is no azdo-cli, but we do have:

https://github.com/git-ecosystem/git-credential-manager

Which I think is the same thing that powers the GitHub experience, I think.  From their landing page, it looks like AzDo is supported.  I do see there's an open issue saying they could work better on code spaces, so maybe there is some debt, but it feels like if I lived in codespaces day to day and used azure, I would want to install this thing and use it.

Instead of generating our `.sh` credential helper that we then use, could we be using this tool instead?  If so, maybe the thing we should be doing is working with the configured git credential manager to store the PAT we generated for the repo in the manager so that future calls just work?  I could get on board with the idea that if `azd` is generating a repository and creating an access token to it, it should run `store` on the credential manager to store that token for future use.  That feels much more reasonable than injecting our own credential manager.   We could ask the user if they wanted to save the credential for future use with git if we were concerned about doing it silently.

Reply:
Either using git-credential-manager with a better flow or prompting to the user stating explicitly what is happening are both fine approaches by me.

Reply:
Just b/c I love you guys so much, I created another PR with a different (azd only) less aggressive approach here: https://github.com/Azure/azure-dev/pull/2088.